rllib/agents/dqn/r2d2_torch_policy.py

```diff
@@ -0,0 +1,338 @@
+
+    Used in [1] in combination with h_inverse:
+      targets = h(r + gamma * h_inverse(Q^))
+    """
+    return torch.sign(x) * (torch.sqrt(torch.abs(x) + 1.0) - 1.0) + epsilon * x
+
+
+def h_inverse(x, epsilon=1.0):
+    """Inverse if the above h-function, described in the paper [1].
+
+    If x > 0.0:
+    h-1(x) = [2eps * x + (2eps + 1) - sqrt(4eps x + (2eps + 1)^2)] /
+        (2 * eps^2)
+
+    If x < 0.0:
+    h-1(x) = [2eps * x + (2eps + 1) + sqrt(-4eps x + (2eps + 1)^2)] /
+        (2 * eps^2)
+    """
+    two_epsilon = epsilon * 2
+    if_x_pos = (two_epsilon * x + (two_epsilon + 1.0) -
+                torch.sqrt(4.0 * epsilon * x +
+                           (two_epsilon + 1.0)**2)) / (2.0 * epsilon**2)
+    if_x_neg = (two_epsilon * x - (two_epsilon + 1.0) +
+                torch.sqrt(-4.0 * epsilon * x +
+                           (two_epsilon + 1.0)**2)) / (2.0 * epsilon**2)
+    return torch.where(x < 0.0, if_x_neg, if_x_pos)
+
+
+class ComputeTDErrorMixin:
+    """Assign the `compute_td_error` method to the R2D2TorchPolicy
+
+    This allows us to prioritize on the worker side.
+    """
+
+    def __init__(self):
+        def compute_td_error(obs_t, act_t, rew_t, obs_tp1, done_mask,
+                             importance_weights):
+            input_dict = self._lazy_tensor_dict({SampleBatch.CUR_OBS: obs_t})
+            input_dict[SampleBatch.ACTIONS] = act_t
+            input_dict[SampleBatch.REWARDS] = rew_t
+            input_dict[SampleBatch.NEXT_OBS] = obs_tp1
+            input_dict[SampleBatch.DONES] = done_mask
+            input_dict[PRIO_WEIGHTS] = importance_weights
+
+            # Do forward pass on loss to update td error attribute
+            r2d2_loss(self, self.model, None, input_dict)
+
+            return self._td_error
+
+        self.compute_td_error = compute_td_error
+
+
+def adam_optimizer(policy: Policy,
+                   config: TrainerConfigDict) -> "torch.optim.Optimizer":
+    return torch.optim.Adam(
+        policy.q_func_vars, lr=policy.cur_lr, eps=config["adam_epsilon"])
+
+
+def build_q_stats(policy: Policy, batch) -> Dict[str, TensorType]:
+    return dict({
+        "cur_lr": policy.cur_lr,
+    }, **policy._loss_stats)
+
+
+def setup_early_mixins(policy: Policy, obs_space, action_space,
+                       config: TrainerConfigDict) -> None:
+    LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
+
+
+def before_loss_init(policy: Policy, obs_space: gym.spaces.Space,
+                     action_space: gym.spaces.Space,
+                     config: TrainerConfigDict) -> None:
+    ComputeTDErrorMixin.__init__(policy)
+    TargetNetworkMixin.__init__(policy, obs_space, action_space, config)
+    # Move target net to device (this is done automatically for the
+    # policy.model, but not for any other models the policy has).
+    policy.target_q_model = policy.target_q_model.to(policy.device)
+
+
+def compute_q_values(policy: Policy,
```

Comment:
What is the diff between this method and DQN's compute q values?

Reply:
Fixed. Removed from R2D2 and generalized DQN version of this funciton.